tensorstore/kvstore/s3/credentials/cached_credential_provider.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TENSORSTORE_KVSTORE_S3_CACHED_CREDENTIAL_PROVIDER_H
+#define TENSORSTORE_KVSTORE_S3_CACHED_CREDENTIAL_PROVIDER_H
+
+#include "tensorstore/util/result.h"
+#include "tensorstore/kvstore/s3/credentials/aws_credential_provider.h"
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+/// The Cached Credential Provider encapsulates an instance of AwsCredentialProvider
+/// to provide:
+/// 1. Thread-safe credential retrieval
+/// 2. Caching of credentials
+/// 3. Refresh of credentials upon expiry
```

Comment:
So this aligns with `Credential` in the go code

https://github.com/aws/aws-sdk-go/blob/b2738d211a2ba65d26731880f30bfdb6abd80858/aws/credentials/credentials.go#L209-L367

where it proxies a `Provider`. Here, the `AwsCredentialsProvider` is also implemented, mostly so that downstream entities such as the s3 kvstore can take a generic `AwsCredentialsProvider`